Speaker Diarization Scoring

ftshijt opened the pull request:
Add NIST scoring for standard diarization error rate (der)

The results on three models (upstream + downstream):

1. baseline(fbank) + rnn 7.03
2. apc + rnn 7.20
3. wav2vec2 + rnn 4.36

ftshijt commented:
The der scoring needs installation of sctk, will update that in the documentation

sw005320 commented:
I recommend you to use [dscore](https://github.com/nryant/dscore) instead of sctk.
It is more pythonic and also used for the DIHARD challenge official evaluation.
Also, how did you determine `-c 0.25` (collar)?

ftshijt commented:
> I recommend you to use [dscore](https://github.com/nryant/dscore) instead of sctk.
> It is more pythonic and also used for the DIHARD challenge official evaluation.
> Also, how did you determine `-c 0.25` (collar)?

Sure, I will chang the scoring to dscore. I've checked the core of dscore and they seem to use `md-eval` as the core, so the results should be aligned. But it will be definitely better to make it more pythonic.

The `0.25` is default set in EEND. I will turn it into dscore default (0ms) in the update version

sw005320 commented:
The introduction of the collar requires some investigations.
I recommend you to set it 0 if we don’t have such investigations.

sw005320 commented:
I think we need to modify the documentation (basically adding speaker diarization functions as a downstream task). 

@andi611, please update https://github.com/s3prl/s3prl/blob/master/file/S3PRL-interface.png to include the speaker diarization task ;)

andi611 commented:
> I think we need to modify the documentation (basically adding speaker diarization functions as a downstream task).
> 
> @andi611, please update https://github.com/s3prl/s3prl/blob/master/file/S3PRL-interface.png to include the speaker diarization task ;)

Sure thing, I've updated the interface image to include the speaker diarization task, as well as the recently supported upstream models of DeCoAr and PASE. I went ahead and directly pushed the updated image to the master branch.

ftshijt commented:
The updated results with collar 0

baseline(fbank) + rnn 10.95
apc + rnn 12.83
wav2vec2 + rnn 8.06

sw005320 commented:
This is very cool!
Wav2vec2 is effective for a mixture of signals!
Can you share the DER breakdown?

ftshijt commented:
> This is very cool!
> Wav2vec2 is effective for a mixture of signals!
> Can you share the DER breakdown?

These are the details in dscore table:
```
System                               DER    JER    B3-Precision    B3-Recall    B3-F1    GKT(ref, sys)    GKT(sys, ref)    H(ref|sys)    H(sys|ref)     MI    NMI
---------------------------------  -----  -----  --------------  -----------  -------  ---------------  ---------------  ------------  ------------  -----  -----
wav2vec2                            8.06   9.35            0.85         0.85     0.85             0.85             0.85          0.39          0.39  12.07   0.97
baseline                           10.95  11.97            0.81         0.82     0.81             0.82             0.81          0.48          0.49  12.01   0.96
apc                                11.94  12.83            0.80         0.81     0.80             0.81             0.80          0.51          0.52  12.00   0.96
```

Seems that I could not get breakdown DER from dscore, the follows are from NIST-md-eval.pl (dscore is an average over different speakers, while md-eval is a weighted sum, so the number is not exactly the same.)
| Model      | MISSED | FALARM    | SPEAKER ERROR | TOTAL_TIME |
| :---        |    :----:   |          ---: |      ---: |  ---: |
| wav2vec2      | 906.37       | 1617.61   | 231.17 | 35601.36 |
| baseline   | 1223.29       | 2237.71  | 242.49 | 35601.36 |
| apc   | 979.14   | 2604.61      | 248.05 | 35601.36 |

ftshijt commented:
This PR is ready to Merge.

leo19941227 commented:
Hey, 

Since we now shift to use dscore, it seems the README.md for this part might need some minor changes.
Can you help me to update the hackmd?
Thanks!!

ftshijt commented:
> Hey,
> 
> Since we now shift to use dscore, it seems that the README.md for this part might need some minor changes.
> Can you help me to update the hackmd?
> Thanks!!

Changed, please take a look at.

leo19941227 commented:
Hi,

Sorry for the late! I am still busy with other application tasks :(
I will check if this task runnable today (for the evaluation part) and will merge it if no problem is encountered.
Thanks!!

ftshijt commented:
> Hi,
> 
> Sorry for the late! I am still busy with other application tasks :(
> I will check if this task runnable today (for the evaluation part) and will merge it if no problem is encountered.
> Thanks!!

Sounds cool~

leo19941227 commented on downstream/diarization/config.yaml:
```diff
@@ -26,11 +26,15 @@ downstream_expert:
     test_dir: ./downstream/diarization/data/test
+  
+  scorerc:
+    score_dir: ./downstream/diarization/wav2vec2_scoring
```
Is it possible we don't use this extra path and directly dump the h5 file into `expdir` which is an argument of DownstreamExpert (specified in command line with --name or --expdir)?
Since user will like to do the following thing in a shell script:
```
python3 -m train --expdir [exp_path]
python3 -m evaluate -t test -e [exp_path]/best-dev.ckpt  # This will share the expdir with training stage
./score.sh [exp_path]
```

leo19941227 replied:
Also, it will be nice if `score.sh` takes arguments in the above format, which can enable a faster pipeline to evaluate lots of upstream representation.

ftshijt replied:
Definitely make sense. Thanks for the suggestion.

ftshijt commented:
I've refactored the dscore repo to show the details of der computation (error, miss, fa, scored_time). So in our following scoring, we could use this as default (I changed the documentation accordingly)

ftshijt commented:
The final output of the new scoring scheme would be:

```
File                               DER    JER    B3-Precision    B3-Recall    B3-F1    GKT(ref, sys)    GKT(sys, ref)    H(ref|sys)    H(sys|ref)    MI    NMI    SCORED    MISS    FA    ERROR
-------------------------------  -----  -----  --------------  -----------  -------  ---------------  ---------------  ------------  ------------  ----  -----  --------  ------  ----  -------
4077-13754-0001_5142-33396-0065  14.03  14.06            0.72         0.78     0.75             0.39             0.41          0.65          0.55  0.41   0.40      7.77    0.16  0.86     0.07
6930-76324-0027_5683-32879-0011   6.33   6.24            0.86         0.95     0.90             0.57             0.29          0.39          0.17  0.23   0.46     15.95    0.24  0.74     0.03
*** OVERALL ***                   8.85  10.15            0.81         0.89     0.85             0.80             0.68          0.48          0.31  1.24   0.76     23.72    0.40  1.60     0.10
```

leo19941227 commented:
Got it! This is wonderful!
I will test this branch by tomorrow (or today if possible).

leo19941227 commented:
Hey!

I have checked the scoring script and it works great.
Thanks very much! I am merging this PR.

leo19941227 commented:
I have made a minor change just for a better ux in my sense [here](https://github.com/s3prl/s3prl/commit/7649c4303e664b4d750d1fe845c6bcb1f690b626#diff-a32aa8358c08e9681ba490bc072cf6bfb7f692cbbe6d8f68faa65002965bfb25R11)
Please let me know if it is inappropriate. Thanks!

ftshijt commented:
> I have made a minor change just for a better ux in my sense [here](https://github.com/s3prl/s3prl/commit/7649c4303e664b4d750d1fe845c6bcb1f690b626#diff-a32aa8358c08e9681ba490bc072cf6bfb7f692cbbe6d8f68faa65002965bfb25R11)
> Please let me know if it is inappropriate. Thanks!

Looks cool!
